numba/parfors/parfor.py

```diff
@@ -4064,28 +4078,198 @@ def remove_duplicate_definitions(blocks, nameset):
+        # that we can't fuse.
+        return index_positions[npsize] != new_position
+
+    def check_index(stmt_index, array_accessed):
+        """Looks at the indices of a getitem or setitem to see if there
+           is a reason that they would prevent fusion.
+           Returns True if fusion should be prohibited, False otherwise.
+        """
+        if isinstance(stmt_index, ir.Var):
+            # If the array is 2+ dimensions then the index should be a tuple.
+            if isinstance(typemap[stmt_index.name], types.BaseTuple):
+                # Get how the index tuple is constructed.
+                fbs_res = guard(find_build_sequence, func_ir, stmt_index)
+                if fbs_res is not None:
+                    ind_seq, ind_op = fbs_res
+                    # If any indices are derived from an index is used then
+                    # return True to say we can't fuse.
+                    if (all([x.name in indices or
+                        x.name not in derived_from_indices for x in ind_seq])):
+                        # Get position in index tuple where parfor indices used.
+                        new_index_positions = [x.name in indices for x in ind_seq]
+                        # Make sure that we aren't accessing a given array with
+                        # different indices in a different order.
+                        if add_check_position(new_index_positions, array_accessed):
+                            return True
+                    else:
+                        # index derived from a parfor index used so no fusion
+                        return True
+                else:
+                    # Don't know how the index tuple is built so
+                    # have to assume fusion can't happen.
+                    return True
+            else:
+                # Should be for 1D arrays.
+                if stmt_index.name in indices:
+                    # Array indexed by a parfor index variable.
+                    # Make sure this 1D access consistent with prior ones.
+                    if add_check_position((True,), array_accessed):
+                        return True
+                elif stmt_index.name in derived_from_indices:
+                    # If we ever index an array with something calculated
+                    # from an index then no fusion.
+                    return True
+                else:
+                    # Some kind of index that isn't a parfor index or
+                    # one derived from one, e.g., a constant.  Make sure
+                    # this is consistent with prior accessed of this array.
+                    if add_check_position((False,), array_accessed):
+                        return True
+        else:
+            # We don't know how to handle non-Var indices so no fusion.
+            return True
+
+        # No reason to reject so indicate safe for fusion.
+        return False
```

Comment:
Just wondering if there's any value in inverting the logic here such that the assumption is that fusion is invalid until proven otherwise? Is there any risk of some unusual case just falling through?

Reply:
Done.